### PR TITLE
[move-2024] Add support for multiline migration, plus comment out friend attributes

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

This fixes a bug where migration would comment out a friend decl but not its attributes, which allowed those attributes to fall onto whichever the next definition was in the file.

While I was at it, my original plan was to grab/comment the entire friend span including its attributes (which didn't work out due to how attributes are processed in expansion), but as an upshot migration now handles multiline fixes as well.

To handle diff output, I roped in the `similar` package which also just looks nicer overall.

## Test Plan 

Updated tests, plus local testing on `sui-framework`.

Note that the expected files have quite a bit of churn due to the change in printing libraries, but the calls to `diff` are still identical, indicating those changes are only in the diff printing.

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/expansion/ast.rs [hunk 1]
```diff
@@ -138,6 +138,7 @@ pub struct ModuleDefinition {
 #[derive(Debug, Clone)]
 pub struct Friend {
     pub attributes: Attributes,
+    pub attr_locs: Vec<Loc>,
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/expansion/ast.rs:141]: Is this not already present in the `Attributes` map?

File: external-crates/move/crates/move-compiler/src/expansion/translate.rs [hunk 2]
```diff
@@ -879,6 +879,11 @@ fn check_visibility_modifiers(
         for (_, _, friend_decl) in friends {
             let loc = friend_decl.loc;
             let diag = if edition == Edition::E2024_MIGRATION {
+                for aloc in &friend_decl.attr_locs {
+                    context
+                        .env()
+                        .add_diag(diag!(Migration::RemoveFriend, (*aloc, friend_msg)));
+                }
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/expansion/translate.rs:886]: Should we just grab the min here?

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp [hunk 3]
```diff
@@ -1,61 +1,55 @@
-error[E13002]: feature is deprecated in specified edition
-  ┌─ tests/move_2024/migration/all_migrations.move:3:5
-  │
-3 │     friend a::n;
-  │     ^^^^^^^^^^^^ 'friend's are deprecated. Remove and replace 'public(friend)' with 'public(package)'
-
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp:6]: Why is this gone?

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/friend_decl_test_only.move [hunk 4]
```diff
@@ -0,0 +1,13 @@
+module a::m {
+
+    #[test_only]
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/migration/friend_decl_test_only.move:3]: We should also test things like 
```
#[
  ext("why"),
 ] friend a::n1
 ```

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/tests/move_2024/migration/friend_decl_test_only.move:3]: And similarly heinous things

File: external-crates/move/crates/move-compiler/src/expansion/ast.rs (hunk 1 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/expansion/ast.rs:141]: No,  the overall spans are gone after `flatten_attributes` happens; only the actual attribute text spans are left. This seemed easier than reworking `flatten_attribute` and all of its callers.

File: external-crates/move/crates/move-compiler/src/expansion/translate.rs (hunk 2 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/expansion/translate.rs:886]: My initial plan was to do that for the whole thing, so:

```
#[test_only]
friend a::m;
```

would become

```
/* #[test_only]
friend a::m; */
```

Unfortunatley this gets a little weird with comments and multiple attributes:

```
/* #[test_only]
#[ext(some_exit)]
// friend for some purpose
friend a::m; */
```

I thought this ultimately looked cleaner when reading the resultant migration code:

```
/* #[test_only] */
/* #[ext(some_exit)] */
// friend for some purpose
/* friend a::m; */
```

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/friend_decl_test_only.move (hunk 4 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/tests/move_2024/migration/friend_decl_test_only.move:3]: I added a number of additional tests.

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp (hunk 3 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp]: it's marked `#[test_only]`

**Review by tnowacki — APPROVED** [review]

**Pull request merged**